### PR TITLE
Add `figure.caption` element

### PR DESCRIPTION
**Pull request by MDLC01**
This small PR adds a `figure.caption` element. I did not open an issue for that beforehand since it is not a complex change anyway.

`figure.caption` serves a similar purpose to `list.item`: it is meant to be used in show rules. For example, you can style all captions in italics with
```typst
#show figure.caption: emph
```

Similarly, I could add an element containing the actual figure (without the caption) if needed.

### CODE REVIEW

File: tests/typ/meta/figure.typ [hunk 1]
```diff
@@ -100,3 +100,19 @@ We can clearly see that @fig-cylinder and
...
+#show figure.caption: emph
+
+#figure("Not italicized", caption: "Italicized")
+
+---
+// Test figure.caption element for specific figure kinds
+#show figure.where(kind: table): it => {
+  show figure.caption: underline
+  it
+}
```
Comment (MDLC01) [inline tests/typ/meta/figure.typ:115]: Is there a simpler way to do that? Something like a show-show rule?

Comment (laurmaedje) [reply tests/typ/meta/figure.typ:115]: There isn't. I explicitly forbade that because it reads a bit weirdly, but maybe we should just support it.

File: crates/typst-library/src/meta/figure.rs [hunk 2]
```diff
@@ -393,6 +398,44 @@ impl FigureElem {
...
+/// )
+/// ```
+///
+/// When using a show rule for figures, you can wrap the full caption in a
+/// `{figure.caption}` element if you want it to be styled as other captions.
+///
+/// Display: Figure Caption
+/// Category: meta
+#[element(Show)]
+pub struct FigureCaption {
```
Comment (laurmaedje) [inline crates/typst-library/src/meta/figure.rs:421]: I feel like the caption should have a few more fields which are populated by the figure's show rule, from which the user can assembly a caption themselves. Like `footnote.entry`.

Comment (laurmaedje) [reply crates/typst-library/src/meta/figure.rs:421]: I more thought of still having it in the figure, but still providing it to the caption as a field of its own (just filled from the figure). But it's a bit tricky.

Comment (MDLC01) [reply crates/typst-library/src/meta/figure.rs:421]: This indeed seems tricky. If I understand well, you want a field that belongs to both a `figure` and its `figure.caption`?

Or maybe the field on caption would be `Smart`, with `auto` inheriting the value from its `figure`, and other values overriding it?

Comment (laurmaedje) [reply crates/typst-library/src/meta/figure.rs:421]: I rather thought the figure would have it as `auto` and the caption doesn't have auto anymore because the figure already resolved it.

Comment (MDLC01) [reply crates/typst-library/src/meta/figure.rs:421]: Oh, you mean, in `FigureElem::show`, construct a `FigureCaption` with the resolved fields? Just want to make sure before implementing it.

Then, `FigureElem::full_caption` would have return a `figure.caption`, making `FigureElem::outline` return a `figure.caption` too. I don't know if this is a problem, but it is a side effect to consider: this means that show rules applying to `figure.caption`s would also apply in the outline.

Basically, if a `figure.caption` constructs its content itself instead of just acting as a proxy to a body that was constructed by its `figure`, we either:
1. Outline figures using with `figure.caption`s, making show rules apply to both the actual caption and in the outline, or
2. Outline figures by constructing separate, distinct, content.

Comment (laurmaedje) [reply crates/typst-library/src/meta/figure.rs:421]: That's what I meant. I haven't thought deeply about it though and don't know whether it would work out well. The outline thing is definitely an important observation. Would be interested what @PgBiel and @Dherse think who've worked on the two things.

Comment (Dherse) [reply crates/typst-library/src/meta/figure.rs:421]: I should also mention that I really like, in general, to have subelements for specific purpose such as a `figure.outline-entry` or `figure.outlined` as a shorthand for creating the outlines as it makes styling them very easy and is imo more composable than the existing technique

Comment (PgBiel) [reply crates/typst-library/src/meta/figure.rs:421]: > Oh, you mean, in `FigureElem::show`, construct a `FigureCaption` with the resolved fields? Just want to make sure before implementing it.

Yes, please refer to `outline.rs` for a sample implementation of the same idea (`OutlineElem`, containing general properties, constructs `OutlineEntry`, with specific per-item properties, for each entry).

> Then, `FigureElem::full_caption` would have return a `figure.caption`, making `FigureElem::outline` return a `figure.caption` too. I don't know if this is a problem, but it is a side effect to consider: this means that show rules applying to `figure.caption`s would also apply in the outline.

I think `full_caption` should only be used for the default, non-`show`-customized caption. The outline would, thus, use this, as _separate and distinct content_.

What I think here is that this `full_caption` method actually belongs to `FigureCaption`, perhaps to its `Show` implementation. Thus, we could just call some sort of `FigureCaption::new` from `FigureElem`, and then call `new_caption.show()` from `FigureElem::outline`, which will return its underlying content (thus avoiding problems with show rules).

>     2. Outline figures by constructing separate, distinct, content.
Therefore, I believe option 2 above is ideal.

Comment (PgBiel) [reply crates/typst-library/src/meta/figure.rs:421]: > I should also mention that I really like, in general, to have subelements for specific purpose such as a `figure.outline-entry` or `figure.outlined` as a shorthand for creating the outlines as it makes styling them very easy and is imo more composable than the existing technique

Not sure if I follow, but if you mean customizing how the figure is displayed in the outline, perhaps this could be redundant, as we already have outline.entry now, and could apply a show rule to entries based on whether `it.element.func() == figure` holds.

Comment (laurmaedje) [reply crates/typst-library/src/meta/figure.rs:421]: This all sounds good. We should separate outline and caption and what @PgBiel says with the Show impl of FigureCaption also makes a lot of sense.

Comment (Dherse) [reply crates/typst-library/src/meta/figure.rs:421]: I had forgotten we have `outline.entry`, see, I wasn't lying when I said I am tired. So, I agree with all of your points so far @PgBiel <3

**Comment by MDLC01** [comment]
I won't be able to work on this until the end of the week. I should be able to implement what was suggested by Sunday or Monday.
Sorry for the inconvenience.

**Comment by laurmaedje** [comment]
Take your time!

**Comment by MDLC01** [comment]
I have tried to implement what was suggested. Let me know if this is not how you imagined it.

**Comment by laurmaedje** [comment]
Just a quick update: I haven't forgotten about this PR. It's just that I experimented with some changes to it locally and then hit some roadblocks. I wanted to take the time to consider the problems carefully and I haven't gotten to doing that yet.

**Comment by laurmaedje** [comment]
Thanks for your work on this!
